x/denommetadata/ibc_middleware.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+
+	"github.com/dymensionxyz/dymension-rdk/x/denommetadata/types"
+)
+
+var _ porttypes.IBCModule = &IBCMiddleware{}
+
+// IBCMiddleware implements the ICS26 callbacks for the transfer middleware
+type IBCMiddleware struct {
+	porttypes.IBCModule
+	keeper DenomMetadataKeeper
+}
+
+type DenomMetadataKeeper interface {
+	HasDenomMetaData(ctx sdk.Context, denom string) bool
+	CreateDenomMetadata(ctx sdk.Context, metadatas ...types.DenomMetadata) error
+}
+
+// NewIBCMiddleware creates a new IBCMiddleware given the keeper and underlying application
+func NewIBCMiddleware(keeper DenomMetadataKeeper, app porttypes.IBCModule) IBCMiddleware {
+	return IBCMiddleware{
+		IBCModule: app,
+		keeper:    keeper,
+	}
+}
+
+// OnRecvPacket registers the denom metadata if it does not exist
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	packetData := new(transfertypes.FungibleTokenPacketData)
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), packetData); err != nil {
+		err = errorsmod.Wrapf(errortypes.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if len(packetData.Memo) == 0 {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	memoData, err := types.ParseMemoData(packetData.Memo)
+	// if the memo is not an object, or does not contain the metadata, we can skip
+	if errors.Is(err, types.ErrMemoUnmarshal) || errors.Is(err, types.ErrMemoDenomMetadataEmpty) {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if memoData.PacketMetadata == nil || memoData.PacketMetadata.DenomMetadata == nil {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	dm := *memoData.PacketMetadata.DenomMetadata
+
+	if err = dm.Validate(); err != nil {
+		return channeltypes.NewErrorAcknowledgement(errortypes.ErrInvalidType)
+	}
+
+	denomTrace := transfertypes.ParseDenomTrace(packetData.Denom)
+	if denomTrace.Path == "" {
+		denomTrace.Path = fmt.Sprintf("%s/%s", packet.GetDestPort(), packet.GetDestChannel())
+	}
+
+	// set the memo back to the original user memo, if there was any
+	packetData.Memo = memoData.UserMemo
+
+	// re-marshal the packet data
+	packet.Data, err = types.ModuleCdc.MarshalJSON(packetData)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if im.keeper.HasDenomMetaData(ctx, denomTrace.IBCDenom()) {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	if err = im.createNewDenom(ctx, dm, denomTrace); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+}
+
+func (im IBCMiddleware) createNewDenom(ctx sdk.Context, denonmMetadata banktypes.Metadata, denomTrace transfertypes.DenomTrace) error {
```

Comment:
typo in denom
Please install a spellchecker on your ide

Reply:
I have it, it complains about `denom` regardless

![Screenshot 2024-05-31 at 16 58 53](https://github.com/dymensionxyz/dymension-rdk/assets/20242145/51a90bcd-45b1-4f49-b70d-b44d538b6a50)

Reply:
you can add it to dict

Reply:
for example I added many common programming abbreviations and dymension terminolog
recv, denom, rollapp etc